### ISSUE TITLE
How do I `yarn tag add react-utilities@2.0.0-beta.1 beta` using Shipit?

### ISSUE THREAD

**Issue by kaelig** [Shopify/shipit-engine#735]
My use case is that I'd like to publish a beta to npm and I don't want it to be installed by default when users run `yarn add react-utilities`.

The way I'd publish a version behind a beta flag is:
- version & publish normally
- `yarn tag add react-utilities@2.0.0-beta.1 beta`

I couldn't find how to do this with Shipit - am I missing something?

**Comment by casperisfine** [comment]
Wouldn't this work?

```yaml
deploy:
  post:
    - yarn tag add react-utilities@2.0.0-beta.1 beta
```

(Of course the version part would need to be dynamic, but the idea is there).

**Comment by kaelig** [comment]
cc @Shopify/front-end-tooling

**Comment by kaelig** [comment]
Thanks @casperisfine!

Ideally we'd be able to specify a tag at "deploy to npm"-time. This is something FED Tooling needs, so it'd be nice to have it as a default feature in Shipit.

See https://docs.npmjs.com/getting-started/using-tags for the spec of this npm feature.

**Comment by casperisfine** [comment]
How would you need to specify it? Isn't that something that is committed inside the repo?

**Comment by ismail-syed** [comment]
Shipit runs `npm publish` [here](https://github.com/Shopify/shipit-engine/blob/7cc062e68785a6c5555d59063f816d1e80dbc468/app/models/shipit/deploy_spec/npm_discovery.rb#L66).

We'll need something like [`npm publish --tag beta`](https://docs.npmjs.com/getting-started/using-tags).

For now, I think we can override shipit's deploy with something similar to that @casperisfine listed above: 

```
deploy:
  post:
    - npm publish --tag beta
```

**Comment by casperisfine** [comment]
What I meant is, how should Shipit know if the `--tag beta` thing should be added or not. Is this something we can infer from the repository, or is it something that only humans can know?

**Comment by ismail-syed** [comment]
Off the top of my head, a few ideas: 
- Shipit could infer it's a beta release if the tag contains `beta` as a keyword. (eg. `v2.0.0-beta.1`)
- We could add a shipit settings called `beta stack`, that releases only betas from within that `stack`/`branch`

**Comment by casperisfine** [comment]
The first one sounds good to me.

**Comment by kaelig** [comment]
I thought I'd share this override which should work for now: https://github.com/Shopify/react-utilities/pull/11/files/d0146c81d8f15130ce5297303fad5156c6a358a0#diff-b92192f38bea57b10aa4ff3a6d282b19

**Comment by kaelig** [comment]
I first coded #738 to look for the `beta` pre-release flag in the version number, however I thought that it was a bit restrictive as it doesn't allow for other flags (alpha, rc…). So I rewrote the PR to publish all versions containing pre-release flags under the `next` dist-tag.

This means we could publish anything such as `1.0.0-alpha`, `1.0.0-beta`, `1.0.0-beta.1`, `1.0.0-rc.1`, `1.0.0-gm`… and they would all be installable as `npm i package@next` or `yarn add package@next`.

Let me know what you think!
